src/inc/msquic.h

```diff
@@ -1712,15 +1713,20 @@ __MsQuicClientAttachProvider(
     __MSQUIC_NMR_CLIENT* Client = (__MSQUIC_NMR_CLIENT*)ClientContext;
     void* ProviderContext;
 
-    #pragma warning(suppress:6387) // _Param_(2) could be '0' - by design.
-    Status =
-        NmrClientAttachProvider(
-            NmrBindingHandle,
-            Client,
-            NULL,
-            &ProviderContext,
-            (const void**)&Client->ProviderDispatch);
-    KeSetEvent(&Client->RegistrationCompleteEvent, IO_NO_INCREMENT, FALSE);
+    if (InterlockedIncrement(&Client->BindingCount) == 1) {
```

Comment:
I'm not sure we want to do this in this PR - the rest of the code looks sketchy if more that one NMR binding occurs.

Reply:
We do this to allow exactly one binding. Did I do it wrong?

Reply:
Ah, no, it will prevent more than one client, you're right. It should decrement the count if the attach fails below, but that's not what I was concerned about.

I was mainly concerned that the provider dispatch could get overwritten, but that's within the if block.

Reply:
oh, right. NmrClientAttachProvider could theorectically fail.

Reply:
Hmm, this is still not bulletproof. We need to hold an exclusive lock while doing the unique binding check and calling NmrClientAttachProvider.

Reply:
doesn't the interlocked increment/decrement provide an implicit mutex?

Reply:
There is a case where we increment the binding count to 1 but then NmrClientAttachProvider fails and on another thread clientAttachProvider is being called before the decrement happens. In this case, no bindings will be formed.

Reply:
Oh I see, I assumed that was by design since the first provider could very possibly have succeeded. IMO it's fine to kick that can down the road until we have dynamic load/unload with multiple providers fully specified and tested.

Reply:
Agreed. We can push that back until the provider actually can fail to attach.